### PR TITLE
Add contribution of bins to the measured parameters

### PR DESCRIPTION
**Pull request by vserret**
Add code to compute the contribution of bins to the measured parameters.
Fix issue https://github.com/igmhub/picca/issues/525

### CODE REVIEW

**Comment by londumas** [comment]
@vserret, thank you for the work. Could you give an example of running the code and some output plots?

**Comment by vserret** [comment]
If you write the following command : 
**effective-bins.py --chi2 chi2-auto-std.ini --params 'ap' 'at' 'bias_LYA' 'beta_LYA' --plot-effective-bins**

With this command, you compute the effective z pour the parameters ap, at, bias_LYA and beta_LYA and you display the bins contribution for these 4 parameters. 

**OUTPUT :** 
Parameter ap
('res = ', 1671.4984694690111, 'den = ', 713.93411979021573)
('<z> = ', 2.3412502962600645)
 
Parameter at
('res = ', 863.47414496935733, 'den = ', 369.83411740860021)
('<z> = ', 2.3347606516663624)
 
Parameter bias_LYA
('res = ', 1809216.5925569055, 'den = ', 775176.33019333449)
('<z> = ', 2.3339419975654754)
 
Parameter beta_LYA
('res = ', 5658.5050287131116, 'den = ', 2423.6874737675539)
('<z> = ', 2.3346677696514746)

**PLOTS :** 

ap

at

bias_lya

beta_lya

**Comment by londumas** [comment]
@vserret, Could you change the colormap to `seismic`: https://matplotlib.org/examples/color/colormaps_reference.html? It goes through white for zero.  What is this file `chi2-auto-std.ini` in the command line?

**Comment by vserret** [comment]
Ok
This is the ini file used in the fitter

File: py/picca/fitter2/effective-bins.py [hunk 1]
```diff
@@ -0,0 +1,186 @@
+#!/usr/bin/env python
+import h5py
+#from picca.fitter2 import chi2,data,xi
```
Comment (londumas) [inline py/picca/fitter2/effective-bins.py]: why commented?

File: py/picca/fitter2/effective-bins.py [hunk 2]
```diff
@@ -0,0 +1,186 @@
...
+        for j in range(z.shape[0]):
+            res += M[i,j]*z[i]
+            den += M[i,j]
+    return res,den
+
+if __name__ == '__main__':
+    parser = argparse.ArgumentParser(formatter_class=argparse.ArgumentDefaultsHelpFormatter)
+
+    parser.add_argument('--chi2-file', type = str, default = None, required=True,
+                        help = 'chi2 input file name')
```
Comment (londumas) [inline py/picca/fitter2/effective-bins.py]: Correct spacing and describe better. Something like "Path to the config 'chi2.ini' file used in fitter2"

File: py/picca/fitter2/effective-bins.py [hunk 3]
```diff
@@ -0,0 +1,186 @@
...
+    
+    parser.add_argument('--params', type=str,default=[], required=False,nargs='*',
+        help='list of the fitted parameters')
```
Comment (londumas) [inline py/picca/fitter2/effective-bins.py]: List

File: py/picca/fitter2/effective-bins.py [hunk 4]
```diff
@@ -0,0 +1,186 @@
...
+    
+    parser.add_argument('--plot-effective-bins', action='store_true', required=False,
+        help='display an image with the bins involved in the fit of each selected parameter')
```
Comment (londumas) [inline py/picca/fitter2/effective-bins.py]: Display

File: py/picca/fitter2/effective-bins.py [hunk 5]
```diff
@@ -0,0 +1,186 @@
...
+
+    args = parser.parse_args()
+    chi2_file = args.chi2_file
+
+    if not args.params:
```
Comment (londumas) [inline py/picca/fitter2/effective-bins.py]: No need for that, simply add `required=True` for params

File: py/picca/fitter2/effective-bins.py [hunk 6]
```diff
@@ -0,0 +1,186 @@
+#!/usr/bin/env python
+import h5py
+#from picca.fitter2 import chi2,data,xi
+from picca.fitter2 import parser as fit_parser
+import sys
+import pylab
```
Comment (londumas) [inline py/picca/fitter2/effective-bins.py]: Let's use `matplotlib.pyplot as plt`. Is it possible?

File: py/picca/fitter2/effective-bins.py [hunk 7]
```diff
@@ -0,0 +1,186 @@
+#!/usr/bin/env python
+import h5py
+#from picca.fitter2 import chi2,data,xi
+from picca.fitter2 import parser as fit_parser
```
Comment (londumas) [inline py/picca/fitter2/effective-bins.py]: picca internal library go at the end of import list, easier to track external and internal library

File: py/picca/fitter2/effective-bins.py [hunk 8]
```diff
@@ -0,0 +1,186 @@
...
+import fitsio
+import copy
+import functools
+import argparse
+from scipy import linalg
+if (sys.version_info > (3, 0)):
+    import configparser as ConfigParser
+else:
+    import ConfigParser
+#from picca import wedgize
```
Comment (londumas) [inline py/picca/fitter2/effective-bins.py]: if not used, remove

File: py/picca/fitter2/effective-bins.py (hunk 1 above)
Comment (vserret) [reply py/picca/fitter2/effective-bins.py]: sorry it was for the test 
I remove it

File: py/picca/fitter2/effective-bins.py (hunk 7 above)
Comment (vserret) [reply py/picca/fitter2/effective-bins.py]: I don't understand

**Comment by londumas** [comment]
@vserret, I modified few things so I can run it with path with variables like `$HOME` and also fixed an issue making the plotter plot only `ap`. Please do a `git pull`

File: py/picca/fitter2/effective-bins.py [hunk 9]
```diff
@@ -0,0 +1,189 @@
...
+
+    ### Computation of the effective bins
+    dm_dp = compute_dm_dp(free_pars,best_fit_pars)
+    apply_mask(dm_dp,ico,z,dic_init)
+
+    for p in args.params:
+        print('Parameter {}'.format(p))
+        M = compute_M(dm_dp[p],ico)
+        res,den = compute_z0(M,z)
+        print('res = {}, den = {}'.format(res,den))
```
Comment (londumas) [inline py/picca/fitter2/effective-bins.py]: @vserret, could you be more verbose on what `res` and `den` mean? Maybe use the full name?

File: py/picca/fitter2/effective-bins.py [hunk 10]
```diff
@@ -0,0 +1,189 @@
...
+    res = 0
+    den = 0
+    for i in range(z.shape[0]):
+        for j in range(z.shape[0]):
+            res += M[i,j]*z[i]
+            den += M[i,j]
+    return res,den
+
+if __name__ == '__main__':
+    parser = argparse.ArgumentParser(formatter_class=argparse.ArgumentDefaultsHelpFormatter)
```
Comment (londumas) [inline py/picca/fitter2/effective-bins.py]: @vserret, please add `description = "<a description of what the code does here....>"`

**Comment by londumas** [comment]
I run it with
```~/Programs/igmhub/picca/py/picca/fitter2/effective-bins.py --chi2-file Correlations/Fit/cf_z_0_10_noMet/chi2.ini --params ap at bias_eta_LYA beta_LYA BB-cf_z_0_10-0-broadband_sky-scale-sky BB-cf_z_0_10-0-broadband_sky-sigma-sky --plot-effective-bins --cor-name cf_z_0_10
```
and
```
~/Programs/igmhub/picca/py/picca/fitter2/effective-bins.py --chi2-file Correlations/Fit/xcf_z_0_10_noMet/chi2.ini --params ap at bias_eta_LYA beta_LYA drp_QSO sigma_velo_lorentz_QSO --plot-effective-bins --cor-name xcf_z_0_10
```

**Comment by londumas** [comment]
@vserret, looks good thanks for the work.

**Comment by londumas** [comment]
@vserret, can I merge?

**Comment by vserret** [comment]
Yes sir !

**Comment by londumas** [comment]
@vserret, me lady thank you for the work. Merged.

**Pull request merged**